File: longestIncresingSubSequence.py
# Time Complexity : O(nlgn)
# Space Complexity : O(N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : no


# Your code here along with comments explaining your approach

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        #can do reg dp or do dp w/ binarySearch      
        if not nums:
            return 0
        
        dp = [ 0 for i in range(len(nums))]
        length = 0
        dp[length] = nums[0]
        length += 1
        
        for i in range(1, len(nums)):
            if dp[length-1] < nums[i]:
                dp[length] = nums[i]
                length += 1
            else:
                index = self.bSearch(dp, nums[i], 0, length-1)
                dp[index] = nums[i]
                
        return length 
    
    def bSearch(self, dp, target, left, right):
        while left <= right:
            mid = left + (right-left)//2
            if target == dp[mid]:
                return mid
            elif dp[mid] > target:
                right = mid-1
            else:
                left = mid + 1
        return left

File: Longestincresingsubsequence.java
// Time Complexity :O(n^2)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums==null || nums.length==0)
            return 0;
        int[] dp=new int[nums.length]; 
        int maxcount=1;
        dp[0]=1;
        for(int i=1;i<nums.length;i++){
            int maxval=0;
            for(int j=0;j<i;j++){ //Finding all the elements smaller then current.
                if(nums[i]>nums[j]){
                    maxval=Math.max(maxval,dp[j]);
                }
            }
            dp[i]=maxval+1;
            maxcount=Math.max(maxcount, dp[i]);
        }

        return maxcount;
        
    }
}